rfcs/20240806-c-api/README.md

```diff
@@ -0,0 +1,126 @@
+Habana. This RFC asks for feedback about aligning the oneCCL APIs to be closer
+to other vendor libraries, since this facilitates integration with frameworks
+and upstreaming to the open source. 
+
+One difference between oneCCL and other vendors communication libraries is that
+all other communication libraries have a C API, while oneCCL has a C++ API.
+This is because oneCCL was designed to integrate with SYCL, which is based on
+C++. One of the goals of oneCCL is to support different hardware and vendors,
+such as Intel Data Center GPU Max Series, Intel Core and Intel Xeon family,
+Intel Gaudi, Nvidia or AMD GPUs, among others. 
+
+[ccl-spec]: https://uxlfoundation.github.io/oneAPI-spec/spec/elements/oneCCL/source/index.html
+[hccl-spec]: https://docs.habana.ai/en/latest/API_Reference_Guides/HCCL_APIs/C_API.html
+[nccl-spec]: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api.html
+[rccl-spec]: https://rocm.docs.amd.com/projects/rccl/en/latest/api-reference/api-library.html#api-library
+
+## Proposal
+
+The proposal is to define a C-like API that aligns with current APIs in other
+communication libraries, while introducing a few changes, as described next:
+
+1. Most APIs are C-based like other communication libraries. C++ data
+   structures are hidden behind handles returned to the user, such as
+   `ccl::stream` and `ccl::comm`. 
+
+2. The API is extended with two C++ API functions to support `sycl::queue`:
+
+   - `onecclResult_t  onecclCreateStream(sycl::queue, &oneccl_stream)`
+   - `onecclResult_t onecclReleaseStream(oneccl_stream)`
+
+   Once the sycl::queue is registered, it is hidden behind the ccl stream
+   handle
+
+3. Add functions to allow users to explicitly control the lifetime of objects,
+   instead of relying on the C++ destructors
+
+   - `onecclResult_t onecclCommFinalize(comm)`
+   - `onecclResult_t onecclCommDestroy(comm)`
+
+4. Drop support for out-of-order SYCL queue and SYCL buffers. The current
+   oneCCL library support out of order SYCL queues, but this feature is not
+   used by the users of the library. In general, the collective operations are
+   submitted to an in-order queue. When out-of order behavior is required,
+   commands are submitted to a different in-order queue, and the two queues are
+   synchronized.
+   
+5. Drop support for SYCL buffers. Only [Unified Shared Memory][usm-example] is
+   supported.
+
+[usm-example]: https://www.intel.com/content/www/us/en/developer/articles/code-sample/dpcpp-usm-code-sample.html
+
+### APIs
+
+The tables below contain the NCCL API, the corresponding new proposed oneCCL
+API, and the current oneCCL API. 
+
+#### APIs related with communicator creation.
+
+| NCCL              | oneCCL (proposed C)          | oneCCL (current, C++)   |
+|-------------------|------------------------------|-------------------------|
+|`cudaError_t`      |`onecclResult_t cudaSetDevice(device)(1)`| N/A          |
```

Comment:
What's the device? We should be much more specific here. I think we don't have to cover this API. My proposal is to handle all device handles in ...Config APIs, so for example. For example passing users device to onecclCommInit would look like:

```c++
onecclCommConfig_t config = ONECCL_COMM_CONFIG_INITIALIZER;
config.sycl.queue = &queue;

onecclCommInitConfig(rank, size, ..., &config);
```

Similar pattern could be applied to `onecclStreamCreate`, so we could get rid of `onecclCreateSyclStream` which looks like a workaround rather than an elegant solution:

```c++
onecclStream_t stream;
onecclStreamConfig_t stream_config = ONECCL_STREAM_SYCL_CONFIG_INITIALIZER;
// or onecclStreamConfig_t config = ONECCL_STREAM_CONFIG_INITIALIZER;
config.sycl.queue = &queue;

onecclStreamCreateConfig(&stream, &config);
```

Reply:
fixed